runtime/pubsub/internal/utils/utils.go

```diff
@@ -1,17 +1,35 @@
 package utils
 
 import (
+	"encoding/json"
 	"errors"
 	"fmt"
 	"math"
 	"reflect"
 	"strconv"
 	"sync"
 	"time"
+
+	"encore.dev/beta/errs"
 )
 
 const AttrTag = "pubsub-attr"
 
+// UnmarshalMessage unmarshals a message into a struct. The message must be a JSON object.
+func UnmarshalMessage[T any](attrs map[string]string, data []byte) (msg T, err error) {
```

Comment:
I moved this from the `NewSubscription` function, so it could be reused by the `testImplementation` as well, without duplicating the code.

That way we _know_ that in tests the real marshal and unmarshal code is being fully executed.